ambry-frontend/src/main/java/com/github/ambry/frontend/PutBlobHandler.java

```diff
@@ -0,0 +1,405 @@
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.frontend;
+
+import com.github.ambry.account.Container;
+import com.github.ambry.commons.BlobId;
+import com.github.ambry.commons.Callback;
+import com.github.ambry.commons.RetainingAsyncWritableChannel;
+import com.github.ambry.config.FrontendConfig;
+import com.github.ambry.messageformat.BlobInfo;
+import com.github.ambry.messageformat.BlobProperties;
+import com.github.ambry.rest.RequestPath;
+import com.github.ambry.rest.RestRequest;
+import com.github.ambry.rest.RestResponseChannel;
+import com.github.ambry.rest.RestServiceErrorCode;
+import com.github.ambry.rest.RestServiceException;
+import com.github.ambry.rest.RestUtils;
+import com.github.ambry.router.ChunkInfo;
+import com.github.ambry.router.PutBlobOptions;
+import com.github.ambry.router.PutBlobOptionsBuilder;
+import com.github.ambry.router.Router;
+import com.github.ambry.utils.Pair;
+import com.github.ambry.utils.Time;
+import com.github.ambry.utils.Utils;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import org.json.JSONObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.github.ambry.frontend.FrontendUtils.*;
+
+
+/**
+ * Handler for put named blob requests. The following request types are handled by {@link PutBlobHandler}:
+ * <h2>Direct uploads</h2>
+ * Direct upload requests treat the body of the request as the content to upload to Ambry. The request path should be
+ * "/named/account_name/container_name/blob_name". In these requests, the blob properties and user metadata are supplied as headers. See
+ * {@link RestUtils#buildBlobProperties(Map)} and {@link RestUtils#buildBlobProperties(Map)} for more details.
+ * <h2>Stitched uploads</h2>
+ * Stitched upload requests allow clients to stitch together previously uploaded data chunks into a single logical blob.
+ * The request path should be "/named/account_name/container_name/blob_name", This request accepts the same headers as direct
+ * upload requests for supplying the blob properties and user metadata of the stitched blob, but, instead of the actual
+ * blob content, accepts a UTF-8 JSON object that includes the signed IDs for the chunks to stitch, and header x-ambry-put-mode should set as "STITCH".
+ * <h3>Request body format</h3>
+ * The body of the request should be a JSON object that conforms to the format described in {@link StitchRequestSerDe}.
+ */
+public class PutBlobHandler {
```

Comment:
There is a massive amount of code here copied from PostBlobHandler.  Please refactor.

Reply:
I've done part of the refactor, will keep working on that.